presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java

```diff
@@ -443,76 +437,68 @@ private PrestoSparkQueryExecution(
-                        taskStatsCollector);
-                return ImmutableList.copyOf(transform(prestoSparkTaskExecutor, tuple -> new Tuple2<>(tuple._1, tuple._2.toMaterializedRow())));
+                        new PrestoSparkTaskInputs(ImmutableMap.of(), ImmutableMap.of(), inputs),
+                        taskStatsCollector,
+                        PrestoSparkSerializedPage.class);
+                return ImmutableList.copyOf(prestoSparkTaskExecutor);
             }
 
-            RddAndMore rootRdd = createRdd(root);
+            RddAndMore<PrestoSparkSerializedPage> rootRdd = createRdd(root, PrestoSparkSerializedPage.class);
             return rootRdd.collectAndDestroyDependencies();
         }
 
-        private RddAndMore createRdd(SubPlan subPlan)
+        private <T> RddAndMore<T> createRdd(SubPlan subPlan, Class<T> outputType)
         {
             ImmutableMap.Builder<PlanFragmentId, JavaPairRDD<MutablePartitionId, PrestoSparkMutableRow>> rddInputs = ImmutableMap.builder();
             ImmutableMap.Builder<PlanFragmentId, Broadcast<List<PrestoSparkSerializedPage>>> broadcastInputs = ImmutableMap.builder();
             ImmutableList.Builder<Broadcast<?>> broadcastDependencies = ImmutableList.builder();
 
             for (SubPlan child : subPlan.getChildren()) {
                 PlanFragment childFragment = child.getFragment();
-                RddAndMore childRdd = createRdd(child);
                 if (childFragment.getPartitioningScheme().getPartitioning().getHandle().equals(FIXED_BROADCAST_DISTRIBUTION)) {
-                    List<Tuple2<MutablePartitionId, PrestoSparkMaterializedRow>> broadcastRows = childRdd.collectAndDestroyDependencies();
-                    // TODO: Transform rows to pages on executors (using `RDD#map` function)
-                    // TODO: Transforming it on coordinator results in 2x memory utilization as both,
-                    // TODO: rows and pages have to be kept in memory all at the same time
-                    Iterator<Page> pagesIterator = transformMaterializedRowsToPages(transform(broadcastRows.iterator(), Tuple2::_2), childFragment.getTypes());
-                    Iterator<PrestoSparkSerializedPage> serializedPagesIterator = transform(transform(pagesIterator, pagesSerde::serialize), PrestoSparkUtils::toPrestoSparkSerializedPage);
-                    List<PrestoSparkSerializedPage> serializedPages = new ArrayList<>();
-                    serializedPagesIterator.forEachRemaining(serializedPages::add);
-                    Broadcast<List<PrestoSparkSerializedPage>> broadcast = sparkContext.broadcast(serializedPages);
+                    RddAndMore<PrestoSparkSerializedPage> childRdd = createRdd(child, PrestoSparkSerializedPage.class);
+                    List<PrestoSparkSerializedPage> broadcastPages = childRdd.collectAndDestroyDependencies().stream()
+                            .map(Tuple2::_2)
+                            .collect(toList());
```

Comment:
nit: why not `toImmutableList()`?

Reply:
Broadcast won't be able to serialize it, Guava is loaded by the Presto class loader